docs/rfc/active/tags/README.md

```diff
@@ -0,0 +1,174 @@
+- Start Date: 2021-02-17
+- RFC PR: (after opening the RFC PR, update this with a link to it and update the file name)
+- Discussion Issue: (GitHub issue this was discussed in before the RFC, if any)
+- Implementation PR(s): (leave this empty)
+
+# Tags
+
+## Summary
+
+Adding general support for tag lowers the bar to add simple metadata to entities. Tags can be used as part of the
+business logic (e.g. sensitivity tag on a data set) or more ad hoc use cases, such as a data analyst's favorite
+dashboard. A general tag implementation will also allow us to define and search based on a new kind relationship, like
+which datasets and ML models are marked to contain PII data.
+
+> needs work
+
+## Basic example
+
+> If the proposal involves a new or changed API, include a basic code example. Omit this section if it's not applicable.
+
+## Motivation
+
+> Why are we doing this? What use cases does it support? What is the expected outcome?
+>
+> Please focus on explaining the motivation so that if this RFC is not accepted, the motivation could be used to develop
+> alternative solutions. In other words, enumerate the constraints you are trying to solve without coupling them too
+> closely to the solution you have in mind.
+
+## Requirements
+
+> What specific requirements does your design need to meet? This should ideally be a bulleted list of items you wish to
+> achieve with your design. This can help everyone involved (including yourself!) make sure your design is robust enough
+> to meet these requirements.
+>
+> Once everyone has agreed upon the set of requirements for your design, we can use this list to review the detailed
+> design.
+
+### Extensibility
+
+> Please also call out extensibility requirements. Is this proposal meant to be extended in the future? Are you adding a
+> new API or set of models that others can build on in later? Please list these concerns here as well.
+
+## Non-Requirements
+
+> Call out things you don't want to discuss in detail during this review here, to help focus the conversation. This can
+> include things you may build in the future based off this design, but don't wish to discuss in detail, in which case
+> it may also be wise to explicitly list that extensibility in your design is a requirement.
+>
+> This list can be high level and not detailed. It is to help focus the conversation on what you want to focus on.
+
+## Detailed design
+
+We want to introduce some new under `datahub/metadata-models/src/main/pegasus/com/linkedin/common/`.
+
+### `Tag` entity
+
+First we create a `TagMetadata` entity, which defines the actual tag-object.
```

Comment:
I'm assuming there should be some `value: string` field somewhere :)

Reply:
I assume you're referring to string value of the tag that, for example, would be shown in the UI?  So the `<name>` in the Urn is not suitable to be parsed for this purpose?

Reply:
So conceptually I think there's a few ways to do this. I don't this proposal fits either right now, but I could be misreading.

1. Treat tags as first class entities (meaning they have an URN that uniquely identifies each tag). If I make the tag `plaisted-favorite` only one "copy" of it exists. Then add some "lineage" or other graph edge between the tag and tagged entities. I don't fully see the value in this, as this A) requires a graph search, which seems overkill, and B) I don't think we'll have any extra information on the tag, besides the string itself? Note that in this case, its probably still a good idea to normalize the data and have a string field for the tag name, even if the tag name is also in the URN. Separates data from identifier :) 
2. Treat tags as aspects of existing entities. In this case they don't have an URN, they aren't entities themselves. For an example of something like this today, look at `Status`. The advantage here is that tags are just part of entities (via aspects), and can be treated that way, meaning you can use the search index here rather than the graph index. If the idea is just to get more search terms for an entity, you probably want things easily accessible in the search index, not the graph index.


Consider something like this:

```
// Not sure if we need Tag or if an array[string] is ok in Tags. This is more future proof; allows to add more
// information a tag if we want easily...
record Tag {
  value: string;
}

/**
 * Aspect to tag entities with custom search terms.
 */
record Tags {
  tags: array[Tag];
}

// Now if we want to be able to tag datasets, lets add it to the list of dataset aspects
typeref DatasetAspect = union[
   // bunch of existing stuff...
   Tags
]

// And if we want it in the search index, lets add it to the search model
record DatasetDocument includes BaseDocument {
   // bunch of existing stuff...
   tags: array[string];
}
```

Granted, I guess downsides to this second approach is that you need to edit every entity definition you want to be able to tag, where as in the first approach you don't need to.


Maybe another way to phrase this is "are tags valuable nodes in the metadata graph, or are they just more pieces of metadata on existing entities?" Maybe flesh out the use cases more (again, total ack this is a WIP), and that question can be more easily answered.

Reply:
Like, lets focus on the DataHuber user story first, and then we can figure out the implementation details :)

Reply:
Got it! We're definitely going for the use case 1 ☝️. I tried to make the case for this approach in the RFC, so it adds little value to repeat that explanation here. Thanks, for the clarification, I will add the `name` field to the `TagMetadata`.

Reply:
The benefit of modeling as a top level entity type is that the same tag is defined once, thus helping achieve semantic consistency. We can support autocomplete on tags, as well as the addition of metadata like description in a single place. 

Think of tags like "hashtags". I should be able to click a tag and see all entities associated with the tag.